File: Graph.java
import java.util.*;

/**
 * A weighted, undirected graph backed by an adjacency-list representation
 * (a Map of Maps), mirroring the Python implementation in algopath.graph.
 */
public class Graph {
    private final Map<String, Map<String, Double>> adjacency = new LinkedHashMap<>();

    public void addNode(String node) {
        adjacency.computeIfAbsent(node, k -> new LinkedHashMap<>());
    }

    public void addEdge(String u, String v, double weight) {
        if (weight < 0) {
            throw new IllegalArgumentException("AlgoPath assumes non-negative edge weights.");
        }
        addNode(u);
        addNode(v);
        adjacency.get(u).put(v, weight);
        adjacency.get(v).put(u, weight);
    }

    public Map<String, Double> neighbors(String node) {
        return adjacency.getOrDefault(node, Collections.emptyMap());
    }

    public Set<String> nodes() {
        return adjacency.keySet();
    }

    public boolean hasNode(String node) {
        return adjacency.containsKey(node);
    }

    public int nodeCount() {
        return adjacency.size();
    }

    public int edgeCount() {
        Set<String> seen = new HashSet<>();
        int count = 0;
        for (var entry : adjacency.entrySet()) {
            for (String v : entry.getValue().keySet()) {
                String key = entry.getKey().compareTo(v) < 0
                        ? entry.getKey() + "|" + v
                        : v + "|" + entry.getKey();
                if (!seen.contains(key)) {
                    seen.add(key);
                    count++;
                }
            }
        }
        return count;
    }

    /** Build the same sample "city road network" used by the Python demo. */
    public static Graph sampleCityGraph() {
        Graph g = new Graph();
        Object[][] edges = {
            {"Hyderabad", "Warangal", 150.0}, {"Hyderabad", "Nizamabad", 175.0},
            {"Hyderabad", "Vijayawada", 275.0}, {"Hyderabad", "Karimnagar", 165.0},
            {"Warangal", "Karimnagar", 70.0}, {"Warangal", "Khammam", 130.0},
            {"Khammam", "Vijayawada", 195.0}, {"Nizamabad", "Karimnagar", 120.0},
            {"Vijayawada", "Guntur", 35.0}, {"Vijayawada", "Visakhapatnam", 350.0},
            {"Guntur", "Nellore", 210.0}, {"Nellore", "Chennai", 175.0},
            {"Visakhapatnam", "Chennai", 800.0}, {"Karimnagar", "Khammam", 210.0},
        };
        for (Object[] e : edges) {
            g.addEdge((String) e[0], (String) e[1], (Double) e[2]);
        }
        return g;
    }

    /** Random connected graph generator, for benchmarking. */
    public static Graph randomGraph(int numNodes, double edgeProbability, long seed) {
        Random rng = new Random(seed);
        Graph g = new Graph();
        List<String> names = new ArrayList<>();
        for (int i = 0; i < numNodes; i++) {
            names.add("N" + i);
            g.addNode("N" + i);
        }
        List<String> shuffled = new ArrayList<>(names);
        Collections.shuffle(shuffled, rng);
        for (int i = 0; i < shuffled.size() - 1; i++) {
            g.addEdge(shuffled.get(i), shuffled.get(i + 1), 1 + rng.nextInt(20));
        }
        for (int i = 0; i < names.size(); i++) {
            for (int j = i + 1; j < names.size(); j++) {
                String a = names.get(i), b = names.get(j);
                if (g.neighbors(a).containsKey(b)) continue;
                if (rng.nextDouble() < edgeProbability) {
                    g.addEdge(a, b, 1 + rng.nextInt(20));
                }
            }
        }
        return g;
    }
}

File: graph.py
"""
Graph data structure for AlgoPath.

Implemented from scratch using an adjacency list (a dict of dicts), rather
than relying on a library graph type, so the underlying representation used
by the search algorithms is fully transparent.
"""

from collections import defaultdict


class Graph:
    """A weighted graph backed by an adjacency-list representation.

    Nodes can be any hashable value (str, int, ...). Edges are stored as
    ``adjacency[u][v] = weight``. By default the graph is undirected, i.e.
    adding an edge (u, v) also adds (v, u).
    """

    def __init__(self, directed: bool = False):
        self.directed = directed
        self.adjacency: dict[str, dict[str, float]] = defaultdict(dict)

    def add_node(self, node: str) -> None:
        # Touching the defaultdict entry is enough to register the node,
        # even if it ends up with no edges.
        _ = self.adjacency[node]

    def add_edge(self, u: str, v: str, weight: float = 1.0) -> None:
        if weight < 0:
            raise ValueError("AlgoPath's algorithms assume non-negative edge weights.")
        self.adjacency[u][v] = weight
        if not self.directed:
            self.adjacency[v][u] = weight
        else:
            # Ensure v is still registered as a node even with no out-edges.
            self.add_node(v)

    def neighbors(self, node: str):
        return self.adjacency[node].items()

    def nodes(self):
        return list(self.adjacency.keys())

    def edges(self):
        seen = set()
        result = []
        for u, nbrs in self.adjacency.items():
            for v, w in nbrs.items():
                key = (u, v) if self.directed else tuple(sorted((u, v)))
                if key not in seen:
                    seen.add(key)
                    result.append((u, v, w))
        return result

    def node_count(self) -> int:
        return len(self.adjacency)

    def edge_count(self) -> int:
        return len(self.edges())

    @classmethod
    def random_graph(cls, num_nodes: int, edge_probability: float = 0.15,
                      min_weight: int = 1, max_weight: int = 20, seed: int | None = None):
        """Generate a random connected weighted graph for benchmarking."""
        import random
        rng = random.Random(seed)
        g = cls(directed=False)
        node_names = [f"N{i}" for i in range(num_nodes)]
        for n in node_names:
            g.add_node(n)

        # First guarantee connectivity with a random spanning chain.
        shuffled = node_names[:]
        rng.shuffle(shuffled)
        for a, b in zip(shuffled, shuffled[1:]):
            g.add_edge(a, b, rng.randint(min_weight, max_weight))

        # Then sprinkle extra random edges.
        for i, a in enumerate(node_names):
            for b in node_names[i + 1:]:
                if a in g.adjacency and b in g.adjacency[a]:
                    continue
                if rng.random() < edge_probability:
                    g.add_edge(a, b, rng.randint(min_weight, max_weight))
        return g
